### PR TITLE
Extend compiler support.

### PR DESCRIPTION
**Pull request by paurkedal**
As you can see, I intend to extend compiler support for `ppx_type` down to 4.02.3. The way I did it, means `ppx_tyre` depends on cppo, which but we could use the custom `List.init` function independent of ocaml version, if you prefer. I needed it also as a `{with-tests}` dependency to support newer compilers.

A intend to make a release after this due to a fix in `ppx_regexp`. I can include `ppx_tyre`, as well, if we want the extended compiler support right away.

### CODE REVIEW

**Comment by paurkedal** [comment]
I pushed another commit to upgrade to AST 4.09 to allow using the PPXes with new features, like the let-operators.  With this, I think it's definitely worth making another `ppx_tyre` release, as well.

File: common/compat.cppo.ml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ *)
+
+#if OCAML_VERSION < (4, 6, 0)
+module List = struct
+  include List
+  let init n f =
+    let rec loop acc i = if i < 0 then acc else loop (f i :: acc) (i - 1) in
+    loop [] (n - 1)
+end
+#endif
```
Comment (Drup) [inline common/compat.cppo.ml]: Well, you don't need cppo here, you can just shadow or use `stdlib-shims`

File: tests/test_regexp.cppo.ml [hunk 2]
```diff
@@ -22,6 +22,16 @@ let mkloc = Loc.mkloc
...
 
+#if OCAML_VERSION < (4, 8, 0)
+let pp_location_error = Location.report_error
+#else
+let pp_location_error = Location.print_report
+#endif
+
+#if OCAML_VERSION < (4, 3, 0)
+type ('a,'b) result = Ok of 'a | Error of 'b
+#endif
```
Comment (Drup) [inline tests/test_regexp.cppo.ml]: Again, either `stdlib-shims` or the `result` package. Not sure why we need that at all though.

File: tests/test_regexp.cppo.ml [hunk 3]
```diff
@@ -22,6 +22,16 @@ let mkloc = Loc.mkloc
 let mknoloc = Loc.mknoloc
 let map_loc f {Loc.txt = x; loc} = {Loc.txt = f x; loc}
 
+#if OCAML_VERSION < (4, 8, 0)
+let pp_location_error = Location.report_error
+#else
+let pp_location_error = Location.print_report
+#endif
```
Comment (Drup) [inline tests/test_regexp.cppo.ml]: The `Location` module is snapshot by `ocaml-migrate-parsetree`, you shouldn't need this.

File: common/compat.cppo.ml (hunk 1 above)
Comment (paurkedal) [reply common/compat.cppo.ml]: Thanks for you opinion, I'll shadow it. I though about `stdlib-shims`, as well, but it only provides a `Stdlib` module containing module alias for the top-level modules as they are in the current standard library.

File: tests/test_regexp.cppo.ml (hunk 2 above)
Comment (paurkedal) [reply tests/test_regexp.cppo.ml]: I'll shadow here as well.  And yes, I didn't need to use `result` in particular here, but using a type like it helped.

File: tests/test_regexp.cppo.ml (hunk 3 above)
Comment (paurkedal) [reply tests/test_regexp.cppo.ml]: This one is more tricky.  I did look for a consistent version of it, but `Migrate_parsetree.Ast_409.Location` is just an alias for the compiler-libs version.  There is a deprecated `Locations` and a `Migrate_parsetree__Locations`.  The latter would work, but it has not been aliased under `Migrate_parsetree`, so I think it should be considered private.

My thoughts on how to handle it:

  1. Make a dummy function and make it simple to replace with the recent version if we need to see the error message. It's just for testing anyway.
  2. Use `Migrate_parsetree__Locations` and find a new solution if it breaks in the future.
  3. Let dune substitute the right version. I think this is no more elegant than using cppo, but it avoids the dependency, if that's the goal.

Do you have other ideas?

Comment (Drup) [reply tests/test_regexp.cppo.ml]: Oh, it's only for testing! Then either 1 or 3, whichever is easiest to understand/maintain.

**Comment by Drup** [comment]
Looks good, Thanks for the release!

**Comment by paurkedal** [comment]
Hopefully soon also in opam: https://github.com/ocaml/opam-repository/pull/15394

**Pull request merged**